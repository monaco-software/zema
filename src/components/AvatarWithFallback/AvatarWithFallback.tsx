import React, { ComponentProps, FC, ReactNode } from 'react';
import { Avatar } from 'grommet';
import { User } from 'grommet-icons';
import { GROMMET_COLORS } from '../App/grommetTheme';

type AvatarProps = ComponentProps<typeof Avatar>;
type AvatarPropsWithoutSize = Omit<AvatarProps, 'size'>;

interface Props extends AvatarPropsWithoutSize {
  size: number;
  url?: string | null;
  placeholderNode?: ReactNode;
}

export const AvatarWithFallback: FC<Props> = ({
  url,
  size,
  placeholderNode,
  ...restProps
}) => {
  const avatarSize = `${size}px`;
  const iconSize = Math.floor(size * 0.4).toString();

  if (url) {
    return (
      <Avatar src={url} size={avatarSize} {...restProps} />
    );
  }

  return (
    <Avatar size={avatarSize} background={GROMMET_COLORS.BRAND} {...restProps}>
      {placeholderNode ?? <User color={GROMMET_COLORS.LIGHT_1} size={iconSize} />}
    </Avatar>
  );
};
