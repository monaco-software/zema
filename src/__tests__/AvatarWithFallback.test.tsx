import React from 'react';
import renderer from 'react-test-renderer';
import { AvatarWithFallback } from '@components/AvatarWithFallback/AvatarWithFallback';

it('<AvatarWithFallback /> should render fallback icon', () => {
  const tree = renderer.create(
    <AvatarWithFallback
      size={64}
    />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('<AvatarWithFallback /> should render image', () => {
  const tree = renderer.create(
    <AvatarWithFallback
      size={64}
      url="some-url"
    />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
