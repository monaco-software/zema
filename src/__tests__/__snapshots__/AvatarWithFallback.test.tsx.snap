// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvatarWithFallback /> should render fallback icon 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 gShNZe StyledAvatar-sc-1suyamb-1"
>
  <svg
    aria-label="User"
    className="StyledIcon-ofa7kd-0 la-dnPZ"
    viewBox="0 0 24 24"
  >
    <path
      d="M8,24 L8,19 M16,24 L16,19 M3,24 L3,19 C3,14.0294373 7.02943725,11 12,11 C16.9705627,11 21,14.0294373 21,19 L21,24 M12,11 C14.7614237,11 17,8.76142375 17,6 C17,3.23857625 14.7614237,1 12,1 C9.23857625,1 7,3.23857625 7,6 C7,8.76142375 9.23857625,11 12,11 Z"
      fill="none"
      stroke="#000"
      strokeWidth="2"
    />
  </svg>
</div>
`;

exports[`<AvatarWithFallback /> should render image 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 beqOLf StyledAvatar-sc-1suyamb-1"
/>
`;
